File: C++/Session 7 - DFS/SPOJ_CAM5.cpp
//  Problem from SPOJ
//  https://www.spoj.com/problems/CAM5/

#include <iostream>
#include <vector>
#include <stack>

using namespace std;

int calc_disjoint_sets(int Ni, vector<vector<int>> graph) {
    int result = 0;
    vector<bool> visited;
    for(int i = 0; i < Ni; i++) {
        visited.push_back(false);
    }
    for (int i = 0; i < Ni; i++) {
        if(visited[i]){
            continue;
        } else {
            stack<int> s;
            s.push(i);
            visited[i] = true;
            while (!s.empty()){
                int u = s.top();
                s.pop();
                for (int j = 0; j < graph[u].size(); j++) {
                    int v = graph[u][j];
                    if (!visited[v]) {
                        visited[v] = true;
                        s.push(v);
                    }
                }
            }
            result++;
        }
    }

    return result;
}

int main()
{
    int t;
    vector<int> results = vector<int>();
    cin >> t;
    for(int i = 0; i < t; i++) {
        int Ni, ei;
        cin >> Ni;
        cin >> ei;
        vector<vector<int>> graph;
        for(int j = 0; j < Ni; j++) {
            graph.push_back(vector<int>());
        }
        for(int j = 0; j < ei; j++) {
            int start, end;
            cin >> start >> end;
            graph[start].push_back(end);
            graph[end].push_back(start);
        }
        results.push_back(calc_disjoint_sets(Ni, graph));
    }
    for(int i = 0; i < results.size() - 1; i++) {
        std::cout << results[i] << std::endl;
    }
    if (results.size() > 0) {
        cout << results[results.size() - 1];
    }
    return 0;
}

File: Blue/Session 07 - DFS/SPOJ_CAM5.py
#  Problem from SPOJ
#  https://www.spoj.com/problems/CAM5/


def calc_disjoint_sets(Ni, graph):
    result = 0
    visited = [False for i in range(Ni)]

    for i in range(Ni):
        if visited[i]:
            continue
        stack = [i]
        visited[i] = True

        while len(stack) > 0:
            u = stack[-1]
            stack.pop()
            for v in graph[u]:
                if not visited[v]:
                    visited[v] = True
                    stack.append(v)
        result += 1

    return result


def solution():
    results = []
    while True:
        new_line = input().strip()
        if new_line:
            t = int(new_line)
            break
    for i in range(t):
        while True:
            new_line = input().strip()
            if new_line:
                Ni = int(new_line)
                break
        ei = int(input())
        graph = [[] for jj in range(Ni)]
        for j in range(ei):
            start, end = map(int, input().split())
            graph[start].append(end)
            graph[end].append(start)

        results.append(calc_disjoint_sets(Ni, graph))

    print(*results, sep='\n')


solution()
